File: src/services/repository.ts
import {
  validateLibrary,
  reviewSchema,
  configSchema,
  DEFAULT_CONFIG,
  type Library,
  type Review,
  type Config,
} from '../models';
import { starterLibrary } from '../seed';

export interface Storage {
  read(path: string): Promise<string | null>;
  write(path: string, content: string): Promise<void>;
  list(path: string): Promise<string[]>;
  remove(path: string): Promise<void>;
}
const json = (v: unknown) => JSON.stringify(v, null, 2) + '\n';
function canonical(value: unknown): string {
  if (Array.isArray(value)) return '[' + value.map(canonical).join(',') + ']';
  if (value && typeof value === 'object')
    return (
      '{' +
      Object.entries(value)
        .sort(([a], [b]) => a.localeCompare(b))
        .map(([k, v]) => JSON.stringify(k) + ':' + canonical(v))
        .join(',') +
      '}'
    );
  return JSON.stringify(value);
}
export function same(a: unknown, b: unknown) {
  return canonical(a) === canonical(b);
}
export class ConflictError extends Error {
  constructor() {
    super('检测到外部修改。请点击“重新加载”，查看最新数据后再保存；本次修改尚未写入。');
  }
}
type Journal = { schema_version: 1; before: Library | null; after: Library; reviews: Review[] };
export class Repository {
  constructor(public storage: Storage) {}
  private async current() {
    const text = await this.storage.read('data/words.json');
    return text === null ? null : validateLibrary(JSON.parse(text));
  }
  private async appendReviews(reviews: Review[]) {
    for (const month of new Set(reviews.map((r) => r.time.slice(0, 7)))) {
      const path = `data/reviews/${month}.jsonl`;
      const text = (await this.storage.read(path)) || '';
      const existing = text
        .split(/\r?\n/)
        .filter(Boolean)
        .map((line) => reviewSchema.parse(JSON.parse(line)));
      const ids = new Set(existing.map((r) => r.id));
      const fresh = reviews.filter((r) => r.time.startsWith(month) && !ids.has(r.id));
      if (fresh.length)
        await this.storage.write(
          path,
          text +
            (text && !text.endsWith('\n') ? '\n' : '') +
            fresh.map((r) => JSON.stringify(r)).join('\n') +
            '\n',
        );
    }
  }
  async recover() {
    const raw = await this.storage.read('data/pending.json');
    if (!raw) return;
    const j = JSON.parse(raw) as Journal;
    if (j.schema_version !== 1 || !Array.isArray(j.reviews))
      throw new Error('存在无法识别的待恢复事务，请保留目录并检查 data/pending.json。');
    if (j.before !== null) validateLibrary(j.before);
    validateLibrary(j.after);
    j.reviews.forEach((r) => reviewSchema.parse(r));
    const disk = await this.current();
    if (same(disk, j.before)) await this.storage.write('data/words.json', json(j.after));
    else if (!same(disk, j.after))
      throw new Error(
        '未完成的保存与磁盘数据冲突。请保留 data/pending.json，使用备份人工核对后恢复。',
      );
    await this.appendReviews(j.reviews);
    await this.storage.remove('data/pending.json');
  }
  async load() {
    await this.recover();
    const library = await this.current();
    if (!library) throw new Error('目录还没有词库，请初始化此目录。');
    const reviews: Review[] = [];
    for (const name of (await this.storage.list('data/reviews'))
      .filter((n) => /^\d{4}-\d{2}\.jsonl$/.test(n))
      .sort()) {
      const raw = (await this.storage.read(`data/reviews/${name}`)) || '';
      for (const [i, line] of raw.split(/\r?\n/).entries())
        if (line.trim()) {
          try {
            reviews.push(reviewSchema.parse(JSON.parse(line)));
          } catch {
            throw new Error(`历史文件 ${name} 第 ${i + 1} 行损坏，请先修复后加载。`);
          }
        }
    }
    const rawConfig = await this.storage.read('data/config.json');
    const config = rawConfig ? configSchema.parse(JSON.parse(rawConfig)) : { ...DEFAULT_CONFIG };
    return { library, reviews, config };
  }
  async initialize() {
    await this.recover();
    if (await this.current()) return this.load();
    if (!(await this.storage.read('data/config.json')))
      await this.storage.write('data/config.json', json(DEFAULT_CONFIG));
    await this.storage.write(
      'data/meta.json',
      json({ schema_version: 1, created_at: new Date().toISOString(), app: 'cet4-vocab' }),
    );
    const library = starterLibrary();
    await this.storage.write(
      'data/pending.json',
      json({ schema_version: 1, before: null, after: library, reviews: [] }),
    );
    await this.recover();
    return this.load();
  }
  async backup(library?: Library) {
    const current = library || (await this.current());
    if (!current) throw new Error('没有可备份的词库。');
    const name = `words_${new Date().toISOString().replace(/[:.]/g, '-')}_${crypto.randomUUID().slice(0, 8)}.json`;
    await this.storage.write(`backups/${name}`, json(current));
    return name;
  }
  async save(
    expected: Library,
    next: Library,
    reviews: Review[] = [],
    source: Library['updated_by'] = 'web',
  ) {
    await this.recover();
    const disk = await this.current();
    if (!same(disk, expected)) throw new ConflictError();
    const after = validateLibrary({
      ...next,
      revision: expected.revision + 1,
      updated_at: new Date().toISOString(),
      updated_by: source,
    });
    reviews.forEach((r) => reviewSchema.parse(r));
    await this.backup(expected);
    if (!same(await this.current(), expected)) throw new ConflictError();
    await this.storage.write(
      'data/pending.json',
      json({ schema_version: 1, before: expected, after, reviews }),
    );
    await this.recover();
    // Backup retention is best effort AFTER a successful commit.
    try {
      const raw = await this.storage.read('data/config.json');
      const limit = raw ? configSchema.parse(JSON.parse(raw)).backup_limit : 30;
      const names = (await this.storage.list('backups'))
        .filter((n) => /^words_.*\.json$/.test(n))
        .sort()
        .reverse();
      for (const name of names.slice(limit)) await this.storage.remove(`backups/${name}`);
    } catch {
      /* A cleanup failure must not make a successful review look unsaved. */
    }
    return after;
  }
  async saveConfig(expected: Config, next: Config) {
    const current = await this.storage.read('data/config.json');
    if (current && !same(configSchema.parse(JSON.parse(current)), expected))
      throw new ConflictError();
    await this.storage.write('data/config.json', json(configSchema.parse(next)));
  }
  async listBackups() {
    return (await this.storage.list('backups'))
      .filter((n) => /^words_.*\.json$/.test(n))
      .sort()
      .reverse();
  }
  async restore(expected: Library, name: string) {
    if (!(await this.listBackups()).includes(name)) throw new Error('备份不存在。');
    const backup = validateLibrary(JSON.parse((await this.storage.read(`backups/${name}`))!));
    return this.save(expected, backup);
  }
}

File: skills/cet4-vocab-coach/scripts/repository.py
"""Local JSON repository with backups, optimistic concurrency and recoverable commits."""
import json
import os
from pathlib import Path
from contextlib import contextmanager
from uuid import uuid4
from copy import deepcopy
from scheduler import now, new_word
from validator import validate, validate_review, validate_config, DEFAULT_CONFIG, ensure


class ConflictError(RuntimeError):
    pass


class Repository:
    def __init__(self, home=None):
        self.home = Path(home or os.environ.get('CET4_VOCAB_HOME') or Path.home() / 'Documents' / 'CET4-Vocab').expanduser().resolve()

    def read_text(self, relative):
        path = self.home / relative
        return path.read_text(encoding='utf-8-sig') if path.exists() else None

    def atomic_text(self, relative, text):
        path = self.home / relative
        path.parent.mkdir(parents=True, exist_ok=True)
        temporary = path.with_name(path.name + '.' + uuid4().hex + '.tmp')
        try:
            with temporary.open('x', encoding='utf-8', newline='\n') as f:
                f.write(text)
                f.flush()
                os.fsync(f.fileno())
            os.replace(temporary, path)
        finally:
            if temporary.exists():
                temporary.unlink()

    def write_json(self, relative, value):
        self.atomic_text(relative, json.dumps(value, ensure_ascii=False, indent=2) + '\n')

    def current(self):
        text = self.read_text('data/words.json')
        return validate(json.loads(text)) if text is not None else None

    @contextmanager
    def lock(self):
        # Serializes Python processes. Browser tabs use Web Locks; cross-client
        # changes are checked by comparing full expected state before committing.
        self.home.mkdir(parents=True, exist_ok=True)
        lock = self.home / '.codex-write.lock'
        try:
            with lock.open('x', encoding='utf-8') as file:
                json.dump({'pid': os.getpid(), 'created_at': now()}, file)
        except FileExistsError:
            raise RuntimeError(f'Another CLI operation is active. If its process has stopped, remove the stale lock: {lock}') from None
        try:
            yield
        finally:
            lock.unlink(missing_ok=True)

    def append_reviews(self, rows):
        for month in sorted({r['time'][:7] for r in rows}):
            relative = f'data/reviews/{month}.jsonl'
            text = self.read_text(relative) or ''
            existing = [validate_review(json.loads(line)) for line in text.splitlines() if line.strip()]
            ids = {r['id'] for r in existing}
            fresh = [r for r in rows if r['time'].startswith(month) and r['id'] not in ids]
            if fresh:
                self.atomic_text(relative, text + ('\n' if text and not text.endswith('\n') else '') + ''.join(json.dumps(r, ensure_ascii=False) + '\n' for r in fresh))

    def recover(self):
        raw = self.read_text('data/pending.json')
        if not raw:
            return
        journal = json.loads(raw)
        ensure(journal.get('schema_version') == 1 and isinstance(journal.get('reviews'), list), 'Invalid pending journal; preserve it and inspect manually')
        if journal['before'] is not None:
            validate(journal['before'])
        validate(journal['after'])
        for r in journal['reviews']:
            validate_review(r)
        disk = self.current()
        if disk == journal['before']:
            self.write_json('data/words.json', journal['after'])
        elif disk != journal['after']:
            raise ConflictError('Pending transaction conflicts with disk. Preserve pending.json and reconcile with a backup.')
        self.append_reviews(journal['reviews'])
        (self.home / 'data/pending.json').unlink()

    def initialize(self):
        with self.lock():
            self.recover()
            if self.current() is not None:
                return self.current()
            for folder in ['data/reviews', 'backups', 'exports', 'logs']:
                (self.home / folder).mkdir(parents=True, exist_ok=True)
            if self.read_text('data/config.json') is None:
                self.write_json('data/config.json', DEFAULT_CONFIG)
            self.write_json('data/meta.json', {'schema_version': 1, 'created_at': now(), 'app': 'cet4-vocab'})
            words = {}
            for seed in json.loads((Path(__file__).parent.parent / 'assets/seed_words.json').read_text(encoding='utf-8')):
                w = new_word(seed['word'], seed['meaning'], seed.get('pos', ''))
                w.update({k: v for k, v in seed.items() if k not in {'meaning', 'pos'}})
                words[w['word']] = w
            library = validate({'schema_version': 1, 'revision': 0, 'updated_at': now(), 'updated_by': 'codex', 'words': words})
            self.write_json('data/pending.json', {'schema_version': 1, 'before': None, 'after': library, 'reviews': []})
            self.recover()
            return library

    def load(self):
        with self.lock():
            self.recover()
            library = self.current()
            ensure(library is not None, 'No words.json found. Run init first.')
            return library

    def config(self):
        text = self.read_text('data/config.json')
        return validate_config(json.loads(text)) if text else deepcopy(DEFAULT_CONFIG)

    def reviews(self):
        rows = []
        for path in sorted((self.home / 'data/reviews').glob('????-??.jsonl')):
            for i, line in enumerate(path.read_text(encoding='utf-8-sig').splitlines()):
                if line.strip():
                    try:
                        rows.append(validate_review(json.loads(line)))
                    except (ValueError, TypeError) as error:
                        raise ValueError(f'{path.name} line {i + 1}: {error}') from error
        return rows

    def backup(self, library=None):
        library = library or self.current()
        ensure(library is not None, 'No library to back up')
        name = 'words_' + now().replace(':', '-').replace('.', '-') + '_' + uuid4().hex[:8] + '.json'
        self.write_json('backups/' + name, library)
        return self.home / 'backups' / name

    def save(self, expected, next_library, reviews=None, source='codex'):
        reviews = reviews or []
        with self.lock():
            self.recover()
            if self.current() != expected:
                raise ConflictError('Revision/content conflict. Reload the library, then reapply only the intended changes.')
            after = deepcopy(next_library)
            after.update(revision=expected['revision'] + 1, updated_at=now(), updated_by=source)
            validate(after)
            for row in reviews:
                validate_review(row)
            self.backup(expected)
            if self.current() != expected:
                raise ConflictError('The library changed while preparing the backup; reload and retry.')
            self.write_json('data/pending.json', {'schema_version': 1, 'before': expected, 'after': after, 'reviews': reviews})
            self.recover()
            try:
                backups = sorted((self.home / 'backups').glob('words_*.json'), reverse=True)
                for path in backups[self.config()['backup_limit']:]:
                    path.unlink()
            except (OSError, ValueError):
                pass
            return after

    def restore(self, name):
        expected = self.load()
        path = Path(name).resolve()
        if not path.exists():
            path = self.home / 'backups' / name
        after = validate(json.loads(path.read_text(encoding='utf-8-sig')))
        return self.save(expected, after)
